sql-plugin/src/main/scala/com/nvidia/spark/rapids/optimizer/FactDimensionJoinReorder.scala

```diff
@@ -0,0 +1,439 @@
+        j += 1
+      }
+    }
+    dims
+  }
+
+  /**
+   * Extracts items of consecutive inner joins and join conditions.
+   * This method works for bushy trees and left/right deep trees.
+   */
+  def extractInnerJoins(plan: LogicalPlan): (TreeShape, Seq[LogicalPlan], ExpressionSet) = {
+
+    var containsLeftJoin = false
+    var containsRightJoin = false
+
+    def _extractInnerJoins(plan: LogicalPlan): (Seq[LogicalPlan], ExpressionSet) = {
+      plan match {
+        case Join(left, right, Inner, Some(cond), JoinHint.NONE) =>
+          if (containsJoin(left)) {
+            containsLeftJoin = true
+          }
+          if (containsJoin(right)) {
+            containsRightJoin = true
+          }
+          val (leftPlans, leftConditions) = _extractInnerJoins(left)
+          val (rightPlans, rightConditions) = _extractInnerJoins(right)
+          (leftPlans ++ rightPlans, leftConditions ++ rightConditions ++
+            splitConjunctivePredicates(cond))
+        case Project(projectList, j@Join(_, _, Inner, Some(_), JoinHint.NONE))
+          if projectList.forall(_.isInstanceOf[Attribute]) =>
+          _extractInnerJoins(j)
+        case _ =>
+          (Seq(plan), ExpressionSet())
+      }
+    }
+
+    val (plans, conds) = _extractInnerJoins(plan)
+
+    val treeShape = if (containsLeftJoin && containsRightJoin) {
+      Bushy
+    } else if (containsLeftJoin) {
+      LeftDeep
+    } else if (containsRightJoin) {
+      RightDeep
+    } else {
+      // did not find any nested joins
+      LeftDeep
+    }
+
+    (treeShape, plans, conds)
+  }
+
+  /** Determine if a plan contains a join operator */
+  private def containsJoin(plan: LogicalPlan): Boolean = {
+    plan match {
+      case _: Join => true
+      case _ => plan.children.exists(containsJoin)
+    }
+  }
+
+}
+
+sealed trait TreeShape
+object LeftDeep extends TreeShape
+object RightDeep extends TreeShape
+object Bushy extends TreeShape
+
+/**
+ * Wrapper for logical plan with size.
+ */
+case class Relation(plan: LogicalPlan, size: Long) extends PredicateHelper {
+  def hasFilter: Boolean = {
+    def hasFilter(plan: LogicalPlan): Boolean = {
+      plan match {
+        case Filter(cond, _) =>
+          // we ignore IsNotNull filters since they exist for all join keys in all relations
+          !splitConjunctivePredicates(cond).forall(_.isInstanceOf[IsNotNull])
+        case _ => plan.children.exists(hasFilter)
+      }
+    }
+    hasFilter(plan)
+  }
+
+  override def toString: String = s"Relation [size=$size]:\n$plan"
+}
+
+object Relation extends Logging {
+
+  def apply(plan: LogicalPlan): Option[Relation] = {
+    getStats(plan).map(stats => Relation(plan, stats.rowCount.get.toLong))
+  }
+
+  /** Get statistics for underlying relation */
+  def getStats(plan: LogicalPlan): Option[Statistics] = {
+    plan match {
+      case p: LogicalRelation =>
+        Some(inferRowCount(p.schema, p.computeStats()))
+      case other if other.children.length == 1 =>
```

Comment:
why only do this when 1 child, can we have > 1 children?

Reply:
I've added some comments explaining the reasoning here